assets/src/edit-story/components/panels/textStyle/font.js

```diff
@@ -54,18 +55,19 @@ function FontControls({ selectedElements, pushUpdate }) {
   const fontSize = getCommonValue(selectedElements, 'fontSize');
 
   const {
-    textInfo: { fontWeight },
+    textInfo: { fontWeight, isItalic },
```

Comment:
`fontWeight` can be `MULTIPLE_VALUE` here. How does that play out with the font loader?

Reply:
Likewise, `isItalic` being `false` doesn't mean that the entire text field is non-italic. It just means that at least one character is non-italic, the rest can still be.

Reply:
Hm, it appears that all font variants are always loaded, so I'm not fully sure what this is even here for? But it seems to work okay.

Reply:
That could be an issue then. We always load all `@font-face`s. However, the font loading itself is lazy, so we need to proactively force-load the variants that we need. E.g. if it's "Roboto" text which includes bold, non-bold, italic and non-italic sections, in all likelihood, we'll need to ask to load the following font combinations:
1. `normal Roboto`
2. `normal italic Roboto`
3. `bold normal Roboto`
4. `bold italic Roboto`
5. etc.

If a `MULTIPLE_VALUE` is possible, we should actually iterate over all `selectedElements` and just pull these combinations. I see the `selectedElements.map()` there that roughly looks like what I'd expect. But I think you're right, at this is not quite correct. If so, we need to follow up.

Reply:
There's an alternative here, of course. We can actually proactively force load all variants always. That could be a bit of an overkill. It's not uncomment for a font to have 8+ variants. We'd be looking at 2-3M on average to load all variants.

Reply:
My findings were, that we **do** load all the font combinations for the currently active font.

However the browser doesn't actually load the font files before it's used the first time. We just load the style sheet for all the font files, where some of those `@font-face` declarations are irrelevant currently.

But it's not _a lot_ of penalty bytes.

Reply:
@barklund it's not about `@font-face` themselves. This is about the loading of actual fonts. We do have all `@font-face` declared, but we have to deal with the race conditions related to the lazy font loading itself. Currently (without these changes) the typical sequence on font prop change can look like this:

1. Change font family/weight/style combination in the editor that hasn't been loaded before (the actual font file, not just `@font-face`).
2. Browser kicks off font file loading. This is as expected.
3. The user immediately sees fallback font in the canvas. This is a little bit bad.
4. We run text measurements to get the new height of the element based on the fallback font. This is pretty bad.
5. The font loading has finished by the browser. Used styles are re-applied.
6. The user now sees the actual font rendered on the canvas. The end-result is just a flicker.
7. We do not re-trigger measurements, however. The box remains with the wrong height.

The minimal goal for this pull request was to build the API that can be given a set of font combinations and it ensures that they are all loaded and ready to be used by the browser. And I think that part  is solid here.

The maximal goal was to intercept (at least a few) places where font combinations change and use this API to ensure that race conditions do not happen. It's not 100% guaranteed that this is all done properly now and we may reconsider where/how we call this API ideally. But, at the very least, the API itself as developed is correct.

Reply:
Hm okay. But I did notice that #1197 calls for not ever displaying fallbacks. That's not factored in yet, right?

Reply:
Not displaying fallbacks is a low-hanging fruit and only solves point 3 above:

> The user immediately sees fallback font in the canvas. This is a little bit bad.

The critical problem however is point 4:

> We run text measurements to get the new height of the element based on the fallback font. This is pretty bad.